_posts/2020-10-23-deconvolution.md

```diff
@@ -0,0 +1,387 @@
+---
+layout: post
+title: Image Analysis Redux
+tagline: Dask + CuPy + RL
+author: John Kirkham (NVIDIA) and Ben Zaitlen (NVIDIA)
+theme: twitter
+---
+{% include JB/setup %}
+
+Summary
+-------
+
+
+Image Analysis Redux
+--------------------
+
+[Last year](https://blog.dask.org/2019/08/09/image-itk) we experimented with Dask/ITK/Scikit-Image
+to perform large scale image analysis on a stack of 3D images.  Specifically, we looked at `deconvolution`,
+a common method to _deblur_ images.  Now, a year later, we return to these experiments with a better
+understanding of how Dask and CuPy can interact, enhanced serialization methods, and support from the
+open-source community.
+
+Previously we used the [Richardson Lucy (RL)](https://en.wikipedia.org/wiki/Richardson%E2%80%93Lucy_deconvolution) deconvolution
+algorithm from ITK and [Scikit-Image](https://github.com/scikit-image/scikit-image/blob/master/skimage/restoration/deconvolution.py#L329).
+We left off at theorizing how GPUs could potentially help accelerate these workflows.  Starting with Scikit-Image's
+implementation, we naively tried replacing `scipy.signal.convolve` calls with `cupyx.scipy.ndimage.convolve`, and while
```

Comment:
And does `cupyx.scipy.ndimage.convolve` use the direct method or FFT method of convolution?

Reply:
`cupyx.scipy.ndimage.convolve`, I believe, uses the direct method.  

When we started, there were no implementations of [NEP-35](https://numpy.org/neps/nep-0035-array-creation-dispatch-with-array-function.html) nor [NEP-37](https://numpy.org/neps/nep-0037-array-module.html) which makes the first line of the definition for `scipy.signal.convolve`: `np.asarray(...)`, challenging.

Reply:
> which makes the first line of the definition for scipy.signal.convolve: np.asarray(...), challenging.

Whoops, I see now. Sorry 'bout that.

Do you think SciPy's convolve would support an efficient implementation of FFTs on CuPy arrays now? Or would there by other challenges like `np.asarray`?

Reply:
> Whoops, I see now. Sorry 'bout that.

No worries at all

> Do you think SciPy's convolve would support an efficient implementation of FFTs on CuPy arrays now? Or would there by other challenges like np.asarray?

I think this is the biggest stumbling block for interoperability.  This also came up in a dask issue: https://github.com/dask/dask/issues/6440

Reply:
Scott, I'm guessing you are asking has something like PR ( https://github.com/scipy/scipy/pull/5608 ) been done for CuPy? If so, no, but there's an open issue ( https://github.com/cupy/cupy/issues/3378 ). It would certainly be great if something like that were implemented for CuPy 😉

As to leveraging existing work in SciPy, that would be nice. We brought this up last year ( https://github.com/scipy/scipy/issues/10204 ). The response was let's try this with one piece (FFTs) and see how that goes.  This got implemented in PR ( https://github.com/scipy/scipy/pull/10383 ), but requires a bit more hand-holding to dispatch effectively. Another question that would come up is at what level in the code would dispatching yield reasonable performance while also reusing as much code as possible. I don't know the answer there, but maybe someone else does.

In Dask-image Genevieve did a bunch of work on adding dispatching recently ( https://github.com/dask/dask-image/pull/151 ). Though this is a Dask-image solution to this problem. Perhaps if dispatching in SciPy were implemented generally, Dask-image could switch to that solution.

Reply:
> I'm guessing you are asking has something like PR ( scipy/scipy#5608 )

Almost! I started that PR because I ran into similar speed issues with RL deconvolution that I resolved in https://github.com/scikit-image/scikit-image/pull/1792 (then I merged that upstream in https://github.com/scipy/scipy/pull/5608).

> requires a bit more hand-holding to dispatch effectively.

If I'm reading https://github.com/scipy/scipy/pull/10383 and https://github.com/scipy/scipy/issues/12652#issuecomment-667348345 correctly, it sounds like the extra bit of hand-holding would be a context manager. Is that correct? That doesn't sound too bad to me. So then the workflow below would work smoothly?

``` python
import cupy
import scipy.fft
import cupyx.scipy.fft as cp_fft
from skimage.restoration import richardson_lucy

img = cupy.arange(256**2).reshape((256, 256)).astype(float)
psf = cupy.arange(128**2).reshape((128, 128)).astype(float)

with scipy.fft.set_backend(cp_fft):
    out = convolve(img, psf, method="fft")

    # I'm especially interested in this:
    img_deblurred = richardson_lucy(img, psf=psf)
```

Reply:
Ah ok. Well if you'd like to add any missing context to issue ( https://github.com/cupy/cupy/issues/3378 ), that would be great! I think it would be nice if we did similar things in CuPy.

That's a good question. Wasn't sure. So took it for a spin. Guessing `convolve` here means `from scipy.signal import convolve`? Unfortunately there's [`np.asarray` calls]( https://github.com/scipy/scipy/blob/79e64c140a518bc85d9105718301509f43e51444/scipy/signal/signaltools.py#L1375-L1376 ), which trips up CuPy. I'm not totally sure how we would back that out. Maybe we could check if the objects support `__array_function__` and skip that step if so?

Yeah it's not too bad. Though if one is trying to write array agnostic code, one would need to detect array type and then branch based on that. Still not too bad, but it is some extra boiler plate. This might get trickier the more array types that need handling.